primitives/tracing/src/types.rs

```diff
@@ -0,0 +1,648 @@
+		let target = "default".as_bytes().to_vec();
+		WasmMetadata {
+			target,
+			name: Default::default(),
+			level: Default::default(),
+			is_span: true,
+			fields: WasmFields::empty()
+		}
+	}
+}
+
+
+fn decode_field(field: &[u8]) -> &str {
+	core::str::from_utf8(field).unwrap_or_default()
+}
+
+/// Span or Event Attributes
+#[derive(Encode, Decode, Clone, Debug)]
+pub struct WasmEntryAttributes {
+	/// the parent, if directly specified – otherwise assume most inner span
+	pub parent_id: Option<u64>,
+	/// the metadata of the location
+	pub metadata: WasmMetadata,
+	/// the Values provided
+	pub fields: WasmValuesSet,
+}
+
+impl From<&tracing_core::Event<'_>> for WasmEntryAttributes {
+	fn from(evt: &tracing_core::Event<'_>) -> WasmEntryAttributes {
+		let mut fields = WasmValuesSet(Vec::new());
+		evt.record(&mut fields);
+		WasmEntryAttributes {
+			parent_id: evt.parent().map(|id| id.into_u64()),
+			metadata: evt.metadata().into(),
+			fields: fields
+		}
+	}
+}
+
+impl From<&tracing_core::span::Attributes<'_>> for WasmEntryAttributes {
+	fn from(attrs: &tracing_core::span::Attributes<'_>) -> WasmEntryAttributes {
+		let mut fields = WasmValuesSet(Vec::new());
+		attrs.record(&mut fields);
+		WasmEntryAttributes {
+			parent_id: attrs.parent().map(|id| id.into_u64()),
+			metadata: attrs.metadata().into(),
+			fields: fields
+		}
+	}
+}
+
+impl core::default::Default for WasmEntryAttributes {
+	fn default() -> Self {
+		WasmEntryAttributes {
+			parent_id: None,
+			metadata: Default::default(),
+			fields: WasmValuesSet(vec![]),
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+mod std_features {
+
+	use tracing;
+	use tracing_core::{
+		callsite, Level,
+		field::{Value, Field, FieldSet, ValueSet, display},
+		metadata::{Kind, Metadata}
+	};
+	use parking_lot::Mutex;
+	use super::{WasmValue, WasmLevel};
+	use std::collections::hash_map::HashMap;
+	use lazy_static::lazy_static;
+
+	// target, name, fieldset, level, kind
+	type MetadataId = (String, String, Vec<String>, WasmLevel, bool);
+
+	lazy_static! {
+		static ref METADATA: Mutex<HashMap<MetadataId, Metadata<'static>>> = Default::default();
+	}
+
+
+	/// Static entry use for wasm-originated metadata.
+	pub struct WasmCallsite;
+	impl callsite::Callsite for WasmCallsite {
+		fn set_interest(&self, _: tracing_core::Interest) { unimplemented!() }
+		fn metadata(&self) -> &Metadata { unimplemented!() }
+	}
+	static CALLSITE: WasmCallsite =  WasmCallsite;
+
+	/// The identifier we are using to inject the wasm events in the generic `tracing` system
+	pub static WASM_TRACE_FALLBACK: &'static str = "wasm_tracing";
+
+	fn mark_static_str<'a>(inp: &'a str) -> &'static str {
+		unsafe {
+			std::mem::transmute(inp)
+		}
+	}
+
+	fn mark_static_slice<'a>(v : &'a[String]) -> &'static [&'static str] {
+		unsafe {
+			std::mem::transmute(v)
+		}
+	}
+
+	fn get_metadata(
+		target: String,
+		name: String,
+		fields: Vec<String>,
+		level: WasmLevel,
+		is_span: bool
+	) -> &'static Metadata<'static> {
+
```

Comment:
in here we generate a `Metadata` from the input – we are moving the Source-`String`s into the hashmap as the key ... does that keep them around? From `API` I assumed that but need to confirm, I guess.

Anyhow, then using the references to their `&str` to generate a `Metadata` we also store there. if not yet given. As we never remove them from the static HashMap, all should stick around until the end of the process, effectively making them `'static` for the rest of runtime, no?

The benefit over using `box::leak` I see here is that we keep one and the same metadata-struct around and construct more than one per callsite.